packages/subgraph/src/mapping.ts

```diff
@@ -59,113 +57,110 @@ export function handleCrossChainEpochOracle(
 
-  while (rawPayloadData.length > 0) {
-    log.warning("NEW MESSAGE BLOCK {}", [messageBlockCounter.toString()]);
-    // Save raw message
-    let messageBlock = new MessageBlock(
-      [txHash, BigInt.fromI32(messageBlockCounter).toString()].join("-")
-    );
-
-    let tags = getTags(
-      changetype<Bytes>(rawPayloadData.slice(0, PREAMBLE_BIT_LENGTH / 8))
-    );
+    // Save raw message.
+    let messageBlock = new MessageBlock([txHash, i].join("-"));
 
-    rawPayloadData = changetype<Bytes>(
-      rawPayloadData.slice(PREAMBLE_BIT_LENGTH / 8)
-    );
-
-    for (let i = 0; i < tags.length; i++) {
-      log.warning("NEW LOOP", []);
-      log.warning("Payload size now {}", [rawPayloadData.length.toString()]);
-
-      if (rawPayloadData.length == 0) {
-        //rollbackToGlobalState(globalState);
-        //return;
-        break;
-      }
-
-      let bytesRead = executeMessage(
-        tags[i],
-        i,
-        globalState,
-        messageBlock.id,
-        rawPayloadData
-      );
-      rawPayloadData = changetype<Bytes>(rawPayloadData.slice(bytesRead));
-      log.warning("Bytes read: {}", [bytesRead.toString()]);
+    numBytesRead += processMessageBlock(globalState, messageBlock, rawPayloadData);
+    if (numBytesRead == 0) {
+      log.error("Failed to process message block num. {}", [i]);
+      return;
     }
 
-    messageBlock.data = rawPayloadData; // cut it to the amount actually read
-    messageBlock.save();
-    log.warning("END OF MESSAGE BLOCK {}", [messageBlockCounter.toString()]);
-    messageBlockCounter++;
+    log.warning("Finished processing message block num. {}", [i]);
+    numBlocksRead++;
   }
 
   commitToGlobalState(globalState);
 }
 
-// Executes the message and returns the amount of bytes read
-function executeMessage(
-  tag: i32,
-  index: i32,
+export function processMessageBlock(
   globalState: GlobalState,
-  messageBlockID: String,
-  data: Bytes
+  messageBlock: MessageBlock,
+  payload: Bytes
 ): i32 {
+  let numBytesRead = PREAMBLE_BYTE_LENGTH;
+
+  // Read the message block's tags.
+  let preamble = changetype<Bytes>(payload.slice(0, PREAMBLE_BYTE_LENGTH));
+  let tags = getTags(preamble);
+
+  for (let i = 0; i < tags.length; i++) {
+    let messageBytes = changetype<Bytes>(payload.slice(numBytesRead));
+    numBytesRead += processMessage(globalState, messageBlock, i, tags[i], messageBytes);
+  }
+
+  messageBlock.data = changetype<Bytes>(payload.slice(PREAMBLE_BYTE_LENGTH, numBytesRead));
+  messageBlock.save();
+
+  return numBytesRead;
+}
+
+// Finishes decoding the message, executes it, and finally returns the amount
+// of bytes read.
+export function processMessage(
+  globalState: GlobalState,
+  messageBlock: MessageBlock,
+  i: i32,
+  tag: MessageTag,
+  payload: Bytes
+): i32 {
+  log.warning("Processing new message with tag {}", [MessageTag.toString(tag)]);
+  log.warning("The remaining payload is {}", [payload.toHexString()]);
```

Comment:
This bit still makes sense, since the payload sent to this function is sliced beforehand based on the amount of bytes read within `processMessageBlock` (so `processMessageBlock` is aware of the number of bytes read and slices the payload accordingly within its scope), but the function that calls the different `processMessageBlock` functions isn't aware of that, or rather, it doesn't slice the payload sent to `processMessageBlock` beforehand, so in a sense, it will always handle the same first message block within the full payload.